beacon_node/operation_pool/src/lib.rs

```diff
@@ -1038,97 +1194,6 @@ mod release_tests {
         assert_eq!(op_pool.num_attestations(), committees.len());
     }
 
-    /// Adding lots of attestations that only intersect pairwise should lead to two aggregate
-    /// attestations.
```

Comment:
Can we try to bring this test back? I don't understand the fundamental reason why it would fail with the new aggregation, but maybe I'm missing something.

Reply:
I think I may have misunderstood what that test was supposed to be
testing. 
I had thought this was about the storing of the attestations which
previously involved greedily aggregating. Based on your feedback I
assume I should go through the clique aggregation process to arrive
at two aggregate attestations. Is this correct?

Reply:
Yeah I think so. The optimal aggregation should be able to find the two aggregates formed by aggregating all the non-overlapping ones. The test might need some minor tweaks though

Reply:
I added the test back but it's failing for now. I'm trying to figure out exactly why. The BK algo takes forever on the same size set of attestations. So I reduced it and it's producing 7739 cliques from a set of 32 attestations. After removing subsets this goes down to 17.

Reply:
This makes more sense now and the way the attestations are currently constructed we should expect there to be more than 2 cliques.  BK produces multiple cliques that when paired with another fully covers the attester set.

Reply:
Usually there are around 16 aggregates per data, but it's random so could be up to 2x more than that quite often.

I agree the slow performance is concerning. It would be good to investigate this case and see if we can get it running a bit quicker. The persistent data structures (instead of lots of vec clones) might be the way to do it. Or it might just be that BK is slow sometimes due to its non-polynomial runtime.

Reply:
I'm not sure if this test case was designed intentionally for this but the number of maximal cliques is much higher than for a graph with random edges. The minimum degree is N-2 since its a path complement graph. This makes things extra slow.

I have tried using the HashTrieSet but the way I've implemented it doesn't improve things on this test case in a noticeable way. 
![Screenshot from 2023-10-29 11-05-49](https://github.com/sigp/lighthouse/assets/104614073/c3cee2a6-1713-47aa-b7c3-53249a71b957)

Reply:
I think it would be cool to see how fast we can get it, but agree we probably shouldn't block on it. I added some suggestions about the persistent data structures here: https://github.com/sigp/lighthouse/pull/4507#discussion_r1375713327

Reply:
@GeemoCandama I spent some time playing with this test and my conclusion is that we need some way to avoid the exponential blow-up from clique enumeration. 

On graphs of ~70 aggregates in the `attestation_pairwise_overlapping` test I saw BK generate millions of cliques without terminating. The bound on the number of cliques is $O(3^{n/3})$, which for 70 aggregates is enormous (135B). For the realistic number of 16 aggregates it's fine, only ~350, but I would prefer that we don't embed an exponential-time algorithm without an escape hatch.


Some ideas I had were:

- Add a `limit` parameter to BK that aborts the algorithm once `limit` cliques have been produced.
- If there are more than `limit` aggregates for a single data just return the best one without calculating any cliques.